build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/renamerTask.ts

```diff
@@ -0,0 +1,28 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { globFn } from "../../../common/utils";
+import { LeafWithFileStatDoneFileTask } from "./leafTask";
+
+/**
+ * This task enables caching of the results of renaming ESM types files, which we do in our build pipeline to work
+ * around a limitation in tsc-multi.
+ *
+ * This implementation is deliberately unintelligent. It assumes that .d.ts files are renamed to .d.mts in the lib
+ * folder. Since renaming files doesn't leave the source files around, this implementation fakes source files by
+ * treating the .d.ts files in the dist folder as the input and the renamed files in the lib folder as the output. This
+ * is safe because any time the output in the dist folder is updated then it's very likely the lib folder would also
+ * need to be updated for other reasons. In other words this shouldn't cause a lot of unnecessary cache invalidation.
```

Comment:
Can we add a comment that this is not sufficient to trigger every time running the task is warranted?

Seems like a better fake would be to use all of the `lib/**/*.mts` files excluding `*.d.mts`.
Alternatively, add smarts to look for files that shouldn't exist and if they do, then trigger task. This task is ideally triggered at least every time a `lib/**/*.d.ts ` exists.

I am also wondering how this system (for rename and overall) deals with files that are deleted. They just sit around and hope they cause no harm?

Reply:
Proposed fix #18896